Source/GSOrderedSet.m

```diff
@@ -0,0 +1,407 @@
+}
+
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState*)state
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
+  //return GSIMapCountByEnumeratingWithStateObjectsCount
+  //            (&map, state, stackbuf, len);
+
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+- (id) objectAtIndexedSubscript: (NSUInteger)index
+{
+  return[self objectAtIndex: index];
+}
+
+- (BOOL) containsObject: (id)anObject
+{
+  NSUInteger i = 0;
+
+  for (i = 0; i < [self count]; i++)
+    {
+      id obj = [self objectAtIndex: i];
+      if([anObject isEqual: obj])
+	{
+	  return YES;
+	}
+    }
+  
+  return NO;
+}
+
+/* Designated initialiser */
+- (id) initWithObjects: (const id*)objs count: (NSUInteger)c
+{
+  NSUInteger i = 0;
+
+  // Reduce C if it doesn't match.  This is apparently how
+  // macOS behaves.
+  if(c > sizeof(objs))
+    {
+      c = sizeof(objs);
+    }
+
+  // Initialize and fill the set.
+  GSIArrayInitWithZoneAndCapacity(&array, [self zone], c);
+  for (i = 0; i < c; i++)
+    {
+      id obj = objs[i];
+      GSIArrayItem item;
+      
+      if (objs[i] == nil)
+	{
+	  DESTROY(self);
+	  [NSException raise: NSInvalidArgumentException
+		      format: @"Tried to init set with nil value"];
+	}
+
+      item.obj = obj;
+      if(![self containsObject: obj])
+	{
+	  GSIArrayAddItem(&array, item);
+	  RETAIN(obj);
+	}
+    }
+  return self;
+}
+
+- (id) initWithObject: (id)obj
+{
+  id objs[] = {obj};
+  
+  self = [self initWithObjects: objs count: 1];
+  if(self == nil)
+    {
+      NSLog(@"Problem initializing with one element");
+    }
+  return self;
+}
+@end
+
+@implementation GSMutableOrderedSet
+
++ (void) initialize
+{
+  if (self == [GSMutableOrderedSet class])
+    {
+      GSObjCAddClassBehavior(self, [GSOrderedSet class]);
+    }
+}
+
+- (void) addObject: (id)anObject
+{
+  GSIArrayItem item;
+
+  if (anObject == nil)
+    {
+      [NSException raise: NSInvalidArgumentException
+		  format: @"Tried to add nil to set"];
+    }
+
+  if([self containsObject: anObject] == NO)
+    {
+      item.obj = anObject;
+      GSIArrayAddItem(&array, item);
+      RETAIN(anObject);
+      _version++;
+    }
+}
+
+- (void) insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
```

Comment:
Why don't we check containsObject: here?

Reply:
Correct.   It should.

Reply:
Done